### PR TITLE
 Improvements in DNS TTL, platform ID and proxy authentication

### PR DESCRIPTION
**Pull request by vveljko**
- Use TTL from DNS response, cache resolved addresses for that
  period
- Identify Windows vs POSIX and C vs C++ (vs Qt) for reporting
  to the server
- Detect authenticating proxy that works incorrectly, mostly the
  kind of broken proxy that keeps asking for the credentials
  over and over again, as-if we're not providing them.

### CODE REVIEW

File: core/pbhttp_digest.h [hunk 1]
```diff
@@ -60,12 +55,26 @@ enum pbhttp_digest_qop {
...
+    pbhtdig_ParameterError,
+    /** Realm in new 'authentication required' message is different from
+        realm previously used
+     */
+    pbhtdig_DifferentConsecutiveRealms,
+    /** Realm in new 'authentication required' message is the same as
+        realm currently in use
+     */
+    pbhtdig_EqualConsecutiveRealms,
+    /** atribute 'realm' is not found yet in 'authentication required' message header
```
Comment (davidnub) [inline core/pbhttp_digest.h]: Nitpick: Typo in comment, should be `attribute`

File: core/pubnub_subscribe_v2.c [hunk 2]
```diff
@@ -136,16 +133,16 @@ enum pubnub_res pbcc_parse_subscribe_v2_response(struct pbcc_context* p)
     if (jonmpOK == jpresult) {
         struct pbjson_elem titel;
```
Comment (davidnub) [inline core/pubnub_subscribe_v2.c]: Nitpick: Should this variable be `title` ?

File: qt/pubnub_qt.cpp [hunk 3]
```diff
@@ -60,6 +61,36 @@ pubnub_qt::~pubnub_qt()
 }
 
 
+static QString GetOsName()
+ {
+ #if defined(Q_OS_ANDROID)
+     return QLatin1String("android");
```
Comment (davidnub) [inline qt/pubnub_qt.cpp]: Nitpick: Capitalize -> `Android`

File: qt/pubnub_qt.cpp [hunk 4]
```diff
@@ -60,6 +61,36 @@ pubnub_qt::~pubnub_qt()
...
+ #elif defined(Q_OS_BLACKBERRY)
+     return QLatin1String("blackberry");
```
Comment (davidnub) [inline qt/pubnub_qt.cpp]: Nitpick: Capitalize -> `Blackberry`

File: qt/pubnub_qt.cpp [hunk 5]
```diff
@@ -60,6 +61,36 @@ pubnub_qt::~pubnub_qt()
...
+ #elif defined(Q_OS_IOS)
+     return QLatin1String("iOS");
+ #elif defined(Q_OS_MACOS)
+     return QLatin1String("MacOS");
+ #elif defined(Q_OS_TVOS)
+     return QLatin1String("Tvos");
```
Comment (davidnub) [inline qt/pubnub_qt.cpp]: Nitpick: Casing -> `tvOS` 

Per Apple's developer page: https://developer.apple.com/

File: qt/pubnub_qt.cpp [hunk 6]
```diff
@@ -60,6 +61,36 @@ pubnub_qt::~pubnub_qt()
...
+static QString GetOsName()
+ {
+ #if defined(Q_OS_ANDROID)
+     return QLatin1String("android");
+ #elif defined(Q_OS_BLACKBERRY)
+     return QLatin1String("blackberry");
+ #elif defined(Q_OS_IOS)
+     return QLatin1String("iOS");
+ #elif defined(Q_OS_MACOS)
+     return QLatin1String("MacOS");
```
Comment (davidnub) [inline qt/pubnub_qt.cpp]: Nitpick: Casing -> `macOS`

Per Apple's developer page: https://developer.apple.com/

File: qt/pubnub_qt.cpp [hunk 7]
```diff
@@ -60,6 +61,36 @@ pubnub_qt::~pubnub_qt()
...
+ #elif defined(Q_OS_TVOS)
+     return QLatin1String("Tvos");
+ #elif defined(Q_OS_WATCHOS)
+     return QLatin1String("Watchos");
```
Comment (davidnub) [inline qt/pubnub_qt.cpp]: Nitpick: Casing -> `watchOS`

Per Apple's developer page: https://developer.apple.com/

File: qt/pubnub_qt.cpp [hunk 8]
```diff
@@ -60,6 +61,36 @@ pubnub_qt::~pubnub_qt()
...
+ #elif defined(Q_OS_WINCE)
+     return QLatin1String("Wince");
```
Comment (davidnub) [inline qt/pubnub_qt.cpp]: Nitpick: Perhaps we should call it `Windows CE` ?

Their docs have stopped shortening it: https://www.microsoft.com/en-us/download/details.aspx?id=14226

File: core/pubnub_subscribe_v2.c (hunk 2 above)
Comment (vveljko) [reply core/pubnub_subscribe_v2.c]: Actually, no, it's a short-form of "_ti_me _t_oken _el_ement". It's a kind of an inside joke, also, there's a small town called "Titel" in Serbia. :)

**Pull request merged**